### PR TITLE
New type of ephemeral keys

### PR DESCRIPTION
**Pull request by acavailhez-stripe**
## Summary
Allowing for ephemeral keys that are not customer

## Motivation
Stripe Issuing will need those

## Testing
Existing ephemeral keys test apply

### CODE REVIEW

File: Stripe/STPEphemeralKey.m [hunk 1]
```diff
@@ -73,14 +79,11 @@ - (BOOL)isEqual:(id)object {
     if (!object || ![object isKindOfClass:self.class]) {
         return NO;
     }
-    return [self isEqualToResourceKey:object];
+    return [self isEqualToEphemeralKey:object];
 }
 
-- (BOOL)isEqualToResourceKey:(STPEphemeralKey *)other {
-    return [self.stripeID isEqualToString:other.stripeID] 
-        && [self.secret isEqualToString:other.secret]
```
Comment (csabol-stripe) [inline Stripe/STPEphemeralKey.m]: why were these removed?

File: Stripe/STPEphemeralKeyManager.m [hunk 2]
```diff
@@ -64,18 +68,44 @@ - (void)handleWillForegroundNotification {
     // eager refreshses to once per hour.
     if (!self.currentKeyIsUnexpired && self.shouldPerformEagerRefresh) {
         self.lastEagerKeyRefresh = [NSDate date];
-        [self.keyProvider createCustomerKeyWithAPIVersion:self.apiVersion completion:^(NSDictionary *jsonResponse, __unused NSError *error) {
-            STPEphemeralKey *key = [STPEphemeralKey decodedObjectFromAPIResponse:jsonResponse];
-            if (key) {
-                self.customerKey = key;
-            }
+        [self getOrCreateKey:^(__unused STPEphemeralKey * _Nullable ephemeralKey, __unused NSError * _Nullable error) {
+
```
Comment (csabol-stripe) [inline Stripe/STPEphemeralKeyManager.m]: It looks like `self.ephemeralKey` gets set in the actual to `getOrCreateKey`. Could we add some comments here explaining why this is called and why we do nothing with the response here?

File: Stripe/STPEphemeralKeyManager.m [hunk 3]
```diff
@@ -64,18 +68,44 @@ - (void)handleWillForegroundNotification {
...
+        STPEphemeralKey *key = [STPEphemeralKey decodedObjectFromAPIResponse:jsonResponse];
+        if (key) {
+            [self.createKeyPromise succeed:key];
+        } else {
+            // the API request failed
+            if (error) {
+                [self.createKeyPromise fail:error];
+            }
+            // the ephemeral key could not be decoded
+            else {
```
Comment (csabol-stripe) [inline Stripe/STPEphemeralKeyManager.m]: style-nit: I prefer to have the `else` on the same line as the closing brace, `} else {`

File: Stripe/STPEphemeralKeyManager.m [hunk 4]
```diff
@@ -64,18 +68,44 @@ - (void)handleWillForegroundNotification {
...
+            }
+        }
+        self.createKeyPromise = nil;
+    };
+    
+    if ([self.keyProvider conformsToProtocol:@protocol(STPCustomerEphemeralKeyProvider)]) {
+        id<STPCustomerEphemeralKeyProvider> provider = self.keyProvider;
+        [provider createCustomerKeyWithAPIVersion:self.apiVersion completion:jsonCompletion];
+    }
+    else if ([self.keyProvider conformsToProtocol:@protocol(STPIssuingCardEphemeralKeyProvider)]) {
```
Comment (csabol-stripe) [inline Stripe/STPEphemeralKeyManager.m]: style-nit

File: Stripe/STPEphemeralKeyManager.m [hunk 5]
```diff
@@ -64,18 +68,44 @@ - (void)handleWillForegroundNotification {
...
-                self.customerKey = key;
-            }
+        [self getOrCreateKey:^(__unused STPEphemeralKey * _Nullable ephemeralKey, __unused NSError * _Nullable error) {
+            
         }];
     }
 }
 
-- (void)getCustomerKey:(STPEphemeralKeyCompletionBlock)completion {
+- (void)createKey {
```
Comment (csabol-stripe) [inline Stripe/STPEphemeralKeyManager.m]: style-nit: (this is a new one I'm trying to enforce so a lot of existing code doesn't follow it). Internal methods should start with an underscore to differentiate from public methods `- (void)_createKey`

File: Stripe/STPEphemeralKeyManager.m [hunk 6]
```diff
@@ -64,18 +68,44 @@ - (void)handleWillForegroundNotification {
     // eager refreshses to once per hour.
```
Comment (fred-stripe) [inline Stripe/STPEphemeralKeyManager.m]: Minor typo:
```    // eager refreshes to once per hour.```

File: Tests/Tests/STPEphemeralKeyManagerTest.m [hunk 7]
```diff
@@ -32,12 +31,16 @@ - (void)setUp {
     self.apiVersion = @"2015-03-03";
 }
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated"
```
Comment (csabol-stripe) [inline Tests/Tests/STPEphemeralKeyManagerTest.m]: what are we using here that's deprecated?

File: Tests/Tests/STPEphemeralKeyManagerTest.m [hunk 8]
```diff
@@ -32,12 +31,16 @@ - (void)setUp {
...
+
 - (id)mockKeyProviderWithKeyResponse:(NSDictionary *)keyResponse {
     XCTestExpectation *exp = [self expectationWithDescription:@"createCustomerKey"];
     id mockKeyProvider = OCMProtocolMock(@protocol(STPEphemeralKeyProvider));
     OCMStub([mockKeyProvider createCustomerKeyWithAPIVersion:[OCMArg isEqual:self.apiVersion]
                                                   completion:[OCMArg any]])
     .andDo(^(NSInvocation *invocation) {
+        [invocation retainArguments];
```
Comment (csabol-stripe) [inline Tests/Tests/STPEphemeralKeyManagerTest.m]: why is this needed?

File: Stripe/STPEphemeralKeyManager.m [hunk 9]
```diff
@@ -64,18 +68,44 @@ - (void)handleWillForegroundNotification {
...
+            [self.createKeyPromise succeed:key];
+        } else {
+            // the API request failed
+            if (error) {
+                [self.createKeyPromise fail:error];
+            }
+            // the ephemeral key could not be decoded
+            else {
+                [self.createKeyPromise fail:[NSError stp_ephemeralKeyDecodingError]];
+                NSAssert(NO, @"Could not parse the ephemeral key response. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api");
```
Comment (fred-stripe) [inline Stripe/STPEphemeralKeyManager.m]: It might be handy to log the type of key here (Issuing vs Customer), not sure if that's available in this context. My thought is the user will contact us when they see that error, and it'd be nice to know which key provider is causing the issue.

File: Stripe/PublicHeaders/STPEphemeralKeyProvider.h [hunk 10]
```diff
@@ -40,7 +39,44 @@ NS_ASSUME_NONNULL_BEGIN
...
  */
 - (void)createCustomerKeyWithAPIVersion:(NSString *)apiVersion completion:(STPJSONResponseCompletionBlock)completion;
+@end
 
+/**
+ You should make your application's API client conform to this interface.
+ It provides a way for Stripe utility classes to request a new ephemeral key from
+ your backend, which it will use to retrieve and update Stripe API objects.
+ */
+@protocol STPIssuingCardEphemeralKeyProvider <NSObject>
```
Comment (csabol-stripe) [inline Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: I'm not sure these need to be two separate protocols. The documentation says that the API client should conform to both of these, which means checking for `keyProvider` protocol conformance will always succeed for both. 

It's not totally clear to me how or why we intend to differentiate between creating customer and card ephemeral keys? Is the intention for the integrator to create two instances of `STPEphemeralKeyManager` and call the right one as needed?

File: Stripe/PublicHeaders/STPCustomerContext.h [hunk 11]
```diff
@@ -35,7 +35,7 @@ NS_ASSUME_NONNULL_BEGIN
  @param keyProvider   The key provider the customer context will use.
  @return the newly-instantiated customer context.
  */
-- (instancetype)initWithKeyProvider:(id<STPEphemeralKeyProvider>)keyProvider;
+- (instancetype)initWithKeyProvider:(id<STPCustomerEphemeralKeyProvider>)keyProvider;
```
Comment (fred-stripe) [inline Stripe/PublicHeaders/STPCustomerContext.h]: Renaming this class means:

1. We're also going to have to update the Examples, the stripe.com docs, and the [generated docs](https://stripe.github.io/stripe-ios/docs/index.html)
2. We'll need coordinate a release between all three.
3. Users will have to update their existing code when they upgrade, even if they're not using Issuing.

I don't think any of this is blockers, just thinking through the user-facing consequences here.

Comment (acavailhez-stripe) [reply Stripe/PublicHeaders/STPCustomerContext.h]: @jack-stripe Just to make sure: is it in line with what you expected?

File: Stripe/STPEphemeralKey.m (hunk 1 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKey.m]: @jack-stripe ?

File: Stripe/STPEphemeralKeyManager.m (hunk 9 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKeyManager.m]: 👍

File: Stripe/STPEphemeralKeyManager.m (hunk 6 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKeyManager.m]: 👍

File: Stripe/STPEphemeralKeyManager.m (hunk 4 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKeyManager.m]: 👍

File: Stripe/STPEphemeralKeyManager.m (hunk 5 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKeyManager.m]: 👍

File: Tests/Tests/STPEphemeralKeyManagerTest.m (hunk 7 above)
Comment (acavailhez-stripe) [reply Tests/Tests/STPEphemeralKeyManagerTest.m]: @jack-stripe ?

File: Tests/Tests/STPEphemeralKeyManagerTest.m (hunk 8 above)
Comment (acavailhez-stripe) [reply Tests/Tests/STPEphemeralKeyManagerTest.m]: There is a bug using `OCMock` that this avoids: https://github.com/erikdoe/ocmock/issues/147

File: Stripe/PublicHeaders/STPEphemeralKeyProvider.h (hunk 10 above)
Comment (acavailhez-stripe) [reply Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: @jack-stripe ?

File: Stripe/STPEphemeralKeyManager.m (hunk 3 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKeyManager.m]: 👍

File: Stripe/PublicHeaders/STPCustomerContext.h (hunk 11 above)
Comment (jack-stripe) [reply Stripe/PublicHeaders/STPCustomerContext.h]: I believe I safely deprecated `STPEphemeralKeyProvider` by changing it to subclass  `STPCustomerEphemeralKeyProvider` - this would trigger a deprecation warning in existing integrations but not a compiler error. Yes, this will mean a docs update though.

File: Stripe/STPEphemeralKey.m (hunk 1 above)
Comment (jack-stripe) [reply Stripe/STPEphemeralKey.m]: I renamed it to `isEqualToEphemeralKey` ("Resource Key" was an early and terrible name for ephemeral keys)

File: Tests/Tests/STPEphemeralKeyManagerTest.m (hunk 7 above)
Comment (jack-stripe) [reply Tests/Tests/STPEphemeralKeyManagerTest.m]: My mistake, I added this accidentally in my initial testing, feel free to remove these ignores

File: Stripe/PublicHeaders/STPEphemeralKeyProvider.h (hunk 10 above)
Comment (jack-stripe) [reply Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: That's the way I imagined it, yes. You might imagine one API that calls `POST https://myapi.com/users/:id/ephemeral_keys` and one that calls `POST https://myapi.com/cards/:id/ephemeral_keys`. Then you might have one API client client-side that calls the first, and another that calls the second.

The other alternative, as you suggest, is to make a single API - `POST https://myapi.com/ephemeral_keys` (with `user` or `card` parameters) and a single API client to talk to both. The issue is that now the method signature needs to change - we need to tell the app what they should be making an ephemeral key for exactly. So the protocol method would have to be something generic like `createKey:forAssociatedStripeObject:` which just feels like turning a thing to figure out at compile-time into a thing to figure out at runtime.

You raise a good point, though, that this breaks if someone's API client conforms to both protocols. Sadly, I think we'll need to add a property to `STPEphemeralKeyManager` to indicate what kind of key it is intended to make (not the worst thing, as it's a private class). Thoughts @csabol-stripe ?

File: Stripe/STPEphemeralKeyManager.m (hunk 2 above)
Comment (acavailhez-stripe) [reply Stripe/STPEphemeralKeyManager.m]: 👍

File: Tests/Tests/STPEphemeralKeyManagerTest.m (hunk 7 above)
Comment (acavailhez-stripe) [reply Tests/Tests/STPEphemeralKeyManagerTest.m]: 👍

File: Tests/Tests/STPEphemeralKeyManagerTest.m (hunk 8 above)
Comment (csabol-stripe) [reply Tests/Tests/STPEphemeralKeyManagerTest.m]: 👍 a comment referencing this would be helpfule

File: Stripe/STPEphemeralKey.m (hunk 1 above)
Comment (csabol-stripe) [reply Stripe/STPEphemeralKey.m]: ah sorry, I mean why was it changed so it only compares the stripeIDs?

File: Stripe/PublicHeaders/STPEphemeralKeyProvider.h (hunk 10 above)
Comment (csabol-stripe) [reply Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: hmm, what about adding a new method to `STPEphemeralKeyManager` so we have `getCustomerKey:` and `getCardKey:` (or similar)? That way we only have one key manager but the requests are still "typed" at compile time. 

Similarly, what are your thoughts about keeping `STPEphemeralKeyProvider` and just adding `createIssuingCardKeyWithAPIVersion:completion:` to the existing protocol (and marking as optional if we think that would be useful?)

File: Tests/Tests/STPEphemeralKeyManagerTest.m (hunk 8 above)
Comment (acavailhez-stripe) [reply Tests/Tests/STPEphemeralKeyManagerTest.m]: Let me add this

File: Stripe/PublicHeaders/STPEphemeralKeyProvider.h (hunk 10 above)
Comment (jack-stripe) [reply Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: I think I'd prefer to keep a property on `STPEphemeralKeyManager`, because I think it'd be a programmer error to call `getCustomerKey:` and then later `getCardKey:` on the same instance. Nice thing is, assuming that property is set in the initializer, we could assert that the given `keyProvider` conforms to the necessary protocol.

I don't feel as strongly about whether or not to split `STPEphemeralKeyProvider` into 2 protocols, but realistically users are going to be implementing only one of `createCustomerKeyWithAPIVersion:completion` or createIssuingCardKeyWithAPIVersion:completion`, so both would have to be optional. In that case, I think it's a little nicer to just have two protocols so that you can enforce at compile-time (instead of runtime) that the necessary methods are implemented.

Comment (jack-stripe) [reply Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: Thoughts @csabol-stripe? Can also discuss in slack on monday : )

Comment (csabol-stripe) [reply Stripe/PublicHeaders/STPEphemeralKeyProvider.h]: closing the loop: @jack-stripe and I met to discuss this today and we are happy moving forward with the patch as is :)

File: Stripe/STPEphemeralKey.m (hunk 1 above)
Comment (csabol-stripe) [reply Stripe/STPEphemeralKey.m]: ping on this @jack-stripe ? (not going to block patch because id equivalence makes sense as sufficient to me but would be useful to know the reasoning

Comment (jack-stripe) [reply Stripe/STPEphemeralKey.m]: Oh, sorry, missed your comment @csabol-stripe!
I honestly don't remember why I changed that. I don't see a world in which `stripeID` is the same but `secret` is different. I'm happy with either implementation.

**Comment by acavailhez-stripe** [comment]
Thanks everyone
